bucko/registry.py

```diff
@@ -189,13 +228,113 @@ def build(self, image):
+        and
+        https://github.com/containers/image/blob/main/docs/signature-protocols.md
+
+        :param str image: image name, eg. "cp/ibm-ceph/ceph-5-rhel8:latest"
+        :returns: list of signatures for this image's manifest sha256 digest.
+                  Note that one manifest digest can have multiple tags.
+        """
+        # Lots of duplicated code here...
+        (repository, reference) = image.split(':', 1)
+        manifests = self.manifest(repository, reference, manifest_list=True)
+        for manifest in manifests['manifests']:
+            platform = manifest['platform']
+            # hard-coding so we only return signatures for one arch...
+            if platform['architecture'] != 'amd64':
+                continue
+            digest = manifest['digest']
+
+        signatures = []
+        if self.lookaside:
+            # Use that URL + /signature-1, 2, 3, etc. until you hit a 404.
+            # The resulting data is the gpg-signed JSON blob.
+            template = posixpath.join(
+                self.lookaside,
+                repository + '@' + digest.replace(':', '='),
+                'signature-{index}'
+            )
+            index = 1
+            MAX_SIGNATURES = 25  # sanity circuit breaker
+            while index < MAX_SIGNATURES:
+                url = template.format(index=index)
+                response = self.session.get(url)
+                if response.ok:
+                    signatures.append(response.content)
+                    index += 1
+                else:
+                    break
+        else:
+            # No /etc/containers/registries.d for this host.
+            # Query the default "extensions" URL on this registry.
+            token = self.tokens[repository]
+            headers = {'Authorization': 'Bearer %s' % token}
+            baseurl = self.baseurl.replace('/v2', '/extensions/v2')
+            url = posixpath.join(baseurl, repository, 'signatures', digest)
+            r = self.session.get(url, headers=headers)
+            if r.status_code == 404:
+                return []
+            r.raise_for_status()
+            data = r.json()
+            for entry in data['signatures']:
+                # "content" is the base64-encoded GPG binary blob.
+                signature = base64.b64decode(entry['content'])
+                signatures.append(signature)
+        return signatures
+
+    def signature_payloads(self, image):
+        """
+        Return the JSON document payloads that GPG has signed.
+
+        Ignores GPG signature validity and simply returns the data.
+
+        :returns: list of dicts, one per signature/ref
+        """
+        # Pipe each binary file thru /usr/bin/gpg to read the contents
+        # on STDOUT and view the signature info on STDERR
+        # Or, run gpg --verify /tmp/myfile to only verify the signature.
+        payloads = []
+        from subprocess import Popen, PIPE
```

Comment:
Is there a reason you're importing here instead of at the top of the file?

Reply:
Whoops. If I recall correctly, in an earlier version of this patch, I was playing around with [pgpy](https://pypi.org/project/PGPy/), and I didn't want to make that a required dependency for Bucko, since I was getting far afield of Bucko's original purpose. Later I learned that I can just shell out to `gpg`. Now that `subprocess` is stdlib, we can move it to the top.